text/0000-add-lifecycle-config.md

```diff
@@ -0,0 +1,95 @@
+# Meta
+[meta]: #meta
+- Name: Add Lifecycle Configuration File
+- Start Date: 2020-12-17
+- Author(s): @Jabrown85
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: "N/A"
+
+# Summary
+[summary]: #summary
+
+In order to allow platforms to have more dynamic configurations, a new configuration file (`lifecycle.toml`) could be loaded by lifecycle binaries.
+
+# Definitions
+[definitions]: #definitions
+
+- Configuration File - New file to hold configuration that is currently achieved by ENV vars or flags. Proposed name `lifecycle.toml`.
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+  - This will allow platform operators to change configuration in workflows previously not easily available to them.
+- What use cases does it support?
+  - Platforms that run all phases in a single kubernetes pod will have the ability to write configuration before and after each phase for the upcoming phases using a volume.
+- What is the expected outcome?
+  - Platforms that use the buildpacks.io published images `buildpacksio/lifecycle` will have a method to configure how lifecycle behaves by writing a file in a configurable location. 
+  - This will open up platform operators to be able to generate a configuration file based on information that may not be available at pod creation time.
+  - Some platforms will move from setting `CNB_*` env vars to producing a `lifecycle.toml`.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Target persona: platform operator, platform implementor
+
+A new `lifecycle.toml` will be read by lifecycle. By default, this path searched will be `/cnb/config/lifecycle.toml`. `CNB_CONFIG_PATH` would be available to override this location.
+
+An example `lifecycle.toml` might be produced by the platform with various inputs (source code, platform specific configuration, etc.):
+```
+[config]
```

Comment:
I wonder if we should try to keep these keys matching the `lifecycle` flags exactly. So:

* `app` instead of `app_dir`
* `order` instead of `order_path`

OTOH, these more closely match the env vars

Reply:
I went ahead and changed the names to be kebab-case, but I kept them matching the env variants for now. I was thinking it might make sense to be able to set some things that maybe aren't available as flags in the future. Like `CNB_REGISTRY_AUTH` isn't a flag, but you still might want to do the work to get that auth setup in the same pod. Similarly, there may be things like lifecycle feature flags or version selection that could be useful to have here that is likely to be available as an env var but may not have a flag name.